net/libfab/libfab.c

```diff
@@ -0,0 +1,2829 @@
+	}
+	
+	M0_LEAVE();
+}
+
+/**
+ * Encodes the descriptor for a (passive) network buffer.
+ */
+static int libfab_bdesc_encode(struct m0_fab__buf *buf)
+{
+	struct m0_fab__bdesc   *fbd;
+	struct fi_rma_iov      *iov;
+	struct m0_net_buf_desc *nbd = &buf->fb_nb->nb_desc;
+	struct m0_net_buffer   *nb = buf->fb_nb;
+	struct m0_fab__tm      *tm = libfab_buf_ma(nb);
+	int                     seg_nr = nb->nb_buffer.ov_vec.v_nr;
+	int                     i;
+	bool                    is_verbs = libfab_is_verbs(tm);
+
+	M0_PRE(seg_nr <= FAB_IOV_MAX);
+
+	nbd->nbd_len = (sizeof(struct m0_fab__bdesc) +
+			(sizeof(struct fi_rma_iov) * seg_nr));
+	nbd->nbd_data = m0_alloc(nbd->nbd_len);
+	if (nbd->nbd_data == NULL)
+		return M0_RC(-ENOMEM);
+
+	fbd = (struct m0_fab__bdesc *)nbd->nbd_data;
+	libfab_ep_pton(&tm->ftm_pep->fep_name, &fbd->fbd_netaddr);
+	fbd->fbd_bufptr = (uint64_t)buf;
+	fbd->fbd_iov_cnt = (uint64_t)seg_nr;
+	iov = (struct fi_rma_iov *)(nbd->nbd_data + 
+				    sizeof(struct m0_fab__bdesc));
+
+	for (i = 0; i < seg_nr; i++) {
+		iov[i].addr = is_verbs ? (uint64_t)nb->nb_buffer.ov_buf[i] : 0;
+		iov[i].key  = fi_mr_key(buf->fb_mr.bm_mr[i]);
+		iov[i].len  = nb->nb_buffer.ov_vec.v_count[i];
+	}
+
+	return M0_RC(0);
+}
+
+/**
+ * Decodes the descriptor of a (passive) network buffer.
+ */
+static void libfab_bdesc_decode(struct m0_fab__buf *fb, 
+				struct m0_fab__ep_name *epname)
+{
+	struct m0_net_buffer   *nb = fb->fb_nb;
+
+	fb->fb_rbd = (struct m0_fab__bdesc *)(nb->nb_desc.nbd_data);
+	fb->fb_riov = (struct fi_rma_iov *)(nb->nb_desc.nbd_data + 
+					    sizeof(struct m0_fab__bdesc));
+	libfab_ep_ntop(fb->fb_rbd->fbd_netaddr, epname);
+	M0_ASSERT(fb->fb_rbd->fbd_iov_cnt <= FAB_IOV_MAX);
+}
+
+/**
+ * Register the buffer with the appropriate access to the domain
+ */
+static int libfab_buf_dom_reg(struct m0_net_buffer *nb, struct fid_domain *dp)
+{
+	struct m0_fab__buf    *fbp = nb->nb_xprt_private;
+	struct m0_fab__buf_mr *mr = &fbp->fb_mr;
+	uint64_t               key;
+	int                    seg_nr = nb->nb_buffer.ov_vec.v_nr;
+	int                    i;
+	int                    ret = FI_SUCCESS;
+	uint32_t               retry_cnt;
+
+	M0_PRE(fbp != NULL && dp != NULL);
+	M0_PRE(seg_nr <= FAB_IOV_MAX);
+
+	if (fbp->fb_dp == dp)
+		return M0_RC(ret);
+
+	for (i = 0; i < seg_nr; i++) {
+		retry_cnt = 20;
+		key = libfab_mr_keygen();
+		
+		ret = fi_mr_reg(dp, nb->nb_buffer.ov_buf[i], 
+				nb->nb_buffer.ov_vec.v_count[i],
+				FAB_MR_ACCESS, FAB_MR_OFFSET, key,
+				FAB_MR_FLAG, &mr->bm_mr[i], NULL);
+
+		if (ret != FI_SUCCESS) {
+			M0_LOG(M0_ERROR, "fi_mr_reg = %d i =%d key = %"PRIu64,
+			       ret, i, key);
+			while(ret != FI_SUCCESS && retry_cnt > 0)
+			{
+				key = libfab_mr_keygen();
+				ret = fi_mr_reg(dp, nb->nb_buffer.ov_buf[i],
+						nb->nb_buffer.ov_vec.v_count[i],
+						FAB_MR_ACCESS, FAB_MR_OFFSET,
+						key, FAB_MR_FLAG, &mr->bm_mr[i],
+						NULL);
+				retry_cnt--;
+			}
+			M0_ASSERT(ret == FI_SUCCESS);
```

Comment:
Do we need to treat this as error and handle it?
Or we can not proceed, and we have to panic?

Reply:
Failure to register the buffer with the domain means that we cannot guarantee the use of this buffer for any network transfer operations. Hence we should assert.